### ISSUE TITLE
Enable to call implementation contract from a proxy contract

### ISSUE THREAD

**Issue by ericleponner** [hashgraph/hedera-mirror-node-explorer#964]
### Problem

This is a follow-up of this [request](https://github.com/hashgraph/hedera-mirror-node-explorer/issues/782#issuecomment-1908955006).

### Solution

There are some ways to:
1) detect that a contract is a proxy 
2) fetch id / address of the implementation contract from proxy bytecode
but these are both heuristic and need to be investigated.

May be a first approach could be to rely on user's knowledge:
user could be asked to provide id/address of the implementation contract before calling implementation methods from the proxy. Feedback is welcome.

**Comment by mshakeg** [comment]
For proxies that respect [EIP-1967](https://eips.ethereum.org/EIPS/eip-1967) the storage slot at which the implementation contract address(or **Logic contract** as referred to in the EIP doc) is stored at `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc`.

For example the proxy contract `0xb177cf10f4382f671e5923c6c06f7b601d5a04e2` has an implementation contract `0x0e44f5cdabefe34d5b729acfa0a79971ffbc0e7e` as returned in the following request's response:

```shell
curl https://mainnet.hashio.io/api \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_getStorageAt","params":["0xb177cf10f4382f671e5923c6c06f7b601d5a04e2", "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc", "latest"],"id":1,"jsonrpc":"2.0"}'
```

```json
{
    "result": "0x0000000000000000000000000e44f5cdabefe34d5b729acfa0a79971ffbc0e7e",
    "jsonrpc": "2.0",
    "id": 1
}
```

Also, the proxy admin contract address storage slot is `0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103` as outlined in EIP-1967. Using the same proxy above the proxy admin is `0x51803f621c5e90011de58b57fd5b7a92e0e39b08` as returned in the following request's response:

```shell
curl https://mainnet.hashio.io/api \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_getStorageAt","params":["0xb177cf10f4382f671e5923c6c06f7b601d5a04e2", "0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103", "latest"],"id":1,"jsonrpc":"2.0"}'
```

```json
{
    "result": "0x00000000000000000000000051803f621c5e90011de58b57fd5b7a92e0e39b08",
    "jsonrpc": "2.0",
    "id": 1
}
```

---

IMU the mirror node explorer is stateless so this might be a feature to implement on sourcify(or hedera's fork of it), i.e. expose an endpoint to determine and register an EIP-1967 compliant proxy and its logic, beacon and admin contracts.

**Comment by mshakeg** [comment]
created [an issue on sourcify](https://github.com/ethereum/sourcify/issues/1336)

**Comment by ericleponner** [comment]
@mshakeg  thanks for the technical insight 👍

**Comment by mshakeg** [comment]
@ericleponner yes, but you'll have to make a mirror node call for every contract to check if it is an ERC-1967 proxy contract which isn't the case for almost all contracts which is a bit wasteful. A cleaner solution imo would be for a backend service(say sourcify) to add some contract metadata for proxy information, then when the mirror node explorer makes a request to sourcify to get a contract that proxy metadata if present is returned, similar to what was described in my sourcify issue, this will probably take a while for the sourcify team to implement, so your approach will work for a quick resolution.

**Comment by mshakeg** [comment]
@ericleponner no it does not, the admin contract updates the implementation address in the proxy contract.

**Comment by mshakeg** [comment]
Hi @ericleponner and @svienot can you also please add auto-verification(and view&write calls) support for minimal clone proxies(provided the implementation is verified)?

The implementation contract for a minimal clone proxy can be extracted as follows:

```shell
curl -X POST -H "Content-Type: application/json" --data '{
    "jsonrpc":"2.0",
    "method":"eth_getCode",
    "params":["0xc0f4d92437d6549ac5ac6544b21dee9903fd4c74", "latest"],
    "id":1
}' https://mainnet.hashio.io/api
```

response:

```
{
    "result": "0x363d3d373d3d3d363d731c3eb3c090607b6b1e598b1722b4c91847ee80825af43d82803e903d91602b57fd5bf3",
    "jsonrpc": "2.0",
    "id": 1
}
```

The address can be extracted from the result as follows:

```
0x363d3d373d3d3d363d73-bebebebebebebebebebebebebebebebebebebebe-5af43d82803e903d91602b57fd5bf3
0x363d3d373d3d3d363d73-1c3eb3c090607b6b1e598b1722b4c91847ee8082-5af43d82803e903d91602b57fd5bf3
```

i.e. `0x1c3eb3c090607b6b1e598b1722b4c91847ee8082` is the implementation address(which I have verified).

https://eips.ethereum.org/EIPS/eip-1167

**Comment by mshakeg** [comment]
@ericleponner @svienot just a friendly nudge on my last reply regarding clone proxies

**Comment by mshakeg** [comment]
@ericleponner I already did [over here](https://github.com/hashgraph/hedera-mirror-node-explorer/issues/964#issuecomment-2457950233)

The proxy is:

> 0xc0f4d92437d6549ac5ac6544b21dee9903fd4c74

The implementation(extracted from the above proxy) is:

> 0x1c3eb3c090607b6b1e598b1722b4c91847ee8082

**Comment by ericleponner** [comment]
@mshakeg
 I apologize  for the duplicate request :/
Thanks again

**Comment by ericleponner** [comment]
@mshakeg
I created #1504 enhancement. Let's follow discussion over there.
